### PR TITLE
Add a median VAF filter function and a filter_fn watermark

### LINKED ISSUE #183: Add watermark for default filter_fn on Cohort initialization
e.g. "Cohort is filtered by abc_filter by default"

### LINKED ISSUE #190: additional_data should automatically create Patient attributes
`p = Patient(..., additional_data={"hello": 1, "goodbye": 2})` should automatically populate `p.hello` and `p.goodbye`.

Perhaps the same for `Sample`s.

### LINKED ISSUE #191: Add frameshift count function
Count only frameshifts.

### PR DESCRIPTION
**Pull request by tavinathanson**
Moving some things from my `data.py` to `Cohorts`.

* Add a function for median VAF.
* Add a `filter_fn` watermark: fixes https://github.com/hammerlab/cohorts/issues/183.

Also in this PR:

* Automatically populate `Patient` object attributes based on `additional_data`: fixes https://github.com/hammerlab/cohorts/issues/190.
* Do the same with `Sample`.
* Add some more functions: fixes https://github.com/hammerlab/cohorts/issues/191.

### CODE REVIEW

**Comment by tavinathanson** [comment]
@jburos added some more stuff, feel free to review at this point.

File: cohorts/utils.py [hunk 1]
```diff
@@ -185,3 +185,10 @@ def get_logger(name, level=logging.INFO):
         logger.handlers = []
     logger.setLevel(level)
     return logger
+
+def set_attributes(obj, additional_data):
+    if additional_data is not None:
+        for key, value in additional_data.items():
+            if hasattr(obj, key):
+                raise ValueError("Key %s in additional_data already exists in this object" % key)
```
Comment (jburos) [inline cohorts/utils.py]: Seems like you might run into this a fair amount, at least at the patient level. E.g. if the additional_data includes a variable like "hla_alleles" or "deceased" which has already been set. Is this the intended behavior?

**Review by jburos — APPROVED** [review]
In general LGTM - put a minor comment but isn't blocking. Lots of useful things in here. 

These could probably use some additional test cases (e.g. for median_vaf_purity & additional_data) but not that pressing.

File: cohorts/utils.py (hunk 1 above)
Comment (tavinathanson) [reply cohorts/utils.py]: Yeah, I think that's a great point and one I was/am a little worried about.

I wanted to error to make sure that we didn't accidentally override important variables, and it doesn't seem too hard for the user to remove or rename the input `DataFrame` to work around the error.

I don't feel strongly, but one option is to just put this in and see what errors we hit?

File: cohorts/utils.py [hunk 2]
```diff
@@ -185,3 +185,10 @@ def get_logger(name, level=logging.INFO):
...
+            setattr(obj, _strip_column_name(key), value)
```
Comment (tavinathanson) [inline cohorts/utils.py]: I should make it clear that this is happening in the docstring.

Comment (tavinathanson) [reply cohorts/utils.py]: @jburos I just updated `set_attributes` to remove the `None` check at the beginning, instead moving that to `Sample` and `Patient`. Also added a docstring.

**Comment by jburos** [comment]
Nice - thanks @tavinathanson ! LGTM. I like the use of strip_column_name. If the conflict happens a lot we may want to catch the error & provide a more descriptive message in the init function for either Patient or Sample. But for now it's probably good as-is.

**Comment by tavinathanson** [comment]
@jburos I had to fix a few failing tests due to e.g. `os` existing in `additional_data` as well as in `Patient` already. I'm fixing by removing that value from the `additional_data` dictionary, which actually seems correct based on the meaning of "additional data". But we'll definitely hit this error in our other cohorts. A simple replacement of e.g. `id=row["id"]` with `id=row.pop("id")` should fix things for all problematic columns in RCC/bladder/etc.

**Pull request merged**